Mathlib/Lean/Data/NameRel.lean

```diff
@@ -0,0 +1,42 @@
+/-
+Copyright (c) 2023 Joachim Breitner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joachim Breitner
+-/
+
+import Lean.Data.NameMap
+import Lean.Declaration
+import Mathlib.Lean.Expr.Basic
+
+/-!
+## A data structure for a relation on names
+-/
+
+open Lean Meta
+
+namespace Lean
+
+/-- A `NameRel` maps names to sets of names -/
+def NameRel := NameMap NameSet
+
+instance : EmptyCollection NameRel :=
+  inferInstanceAs $ EmptyCollection (NameMap NameSet)
+
+instance : Inhabited NameRel :=
+  inferInstanceAs $ Inhabited (NameMap NameSet)
+
+/-- For all names `n` mentioned in the type of the constant `c`, add `c.name` to the set associated
+with `n`. -/
+def NameRel.addDecl (c : ConstantInfo) (m : NameRel) := do
```

Comment:
Now that this is in a generic file, it needs a more specific name. Perhaps `addReferencesFromType`? (Something that indicated in which direction the lookup was happening would be ideal...)

Reply:
Oh, right. This shouldn’t even be in this module, I wasn’t paying close attention; there should just be `NameRel.add : NameRel → Name → Name → NameRel`.